### ISSUE TITLE
invalid memory address or nil pointer dereference

### ISSUE THREAD

**Issue by kjelle** [D4-project/sensor-d4-tls-fingerprinting#7]
Hello.

I testet your program on our x86_64 Centos 7.6.1810 sensor, but after a minute it segfaults:

> panic: runtime error: invalid memory address or nil pointer dereference
> [signal SIGSEGV: segmentation violation code=0x1 addr=0x38 pc=0x5b026e]
>  
> goroutine 1 [running]:
> github.com/D4-project/sensor-d4-tls-fingerprinting/d4tls.(*TLSSession).ja3(0xc4208dbd00, 0x62ffe0)
>         /go/src/github.com/D4-project/sensor-d4-tls-fingerprinting/d4tls/fingerprinter.go:84 +0x3e
> github.com/D4-project/sensor-d4-tls-fingerprinting/d4tls.(*TLSSession).D4Fingerprinting(0xc4208dbd00, 0x64210e, 0x3, 0xc4273b6090)
>         /go/src/github.com/D4-project/sensor-d4-tls-fingerprinting/d4tls/fingerprinter.go:24 +0xb1
> main.(*tcpStream).ReassembledSG(0xc4208dbc20, 0x681ce0, 0xc420012bf8, 0x67d8c0, 0xc425a735e0)
>         /go/src/github.com/D4-project/sensor-d4-tls-fingerprinting/d4-tlsf.go:205 +0x580
> github.com/google/gopacket/reassembly.(*Assembler).sendToConnection(0xc420012b40, 0xc42135ecb0, 0xc42135edc0, 0x67d8c0, 0xc425a735e0, 0xc42135edc0)
>         /go/src/github.com/google/gopacket/reassembly/tcpassembly.go:1099 +0xab
> github.com/google/gopacket/reassembly.(*Assembler).AssembleWithContext(0xc420012b40, 0x1, 0x4, 0x4, 0xc7e7c60a, 0x0, 0xa315d9ac, 0x0, 0xc426ec7400, 0x67d8c0, ...)
>         /go/src/github.com/google/gopacket/reassembly/tcpassembly.go:727 +0x8b0
> main.main()
>         /go/src/github.com/D4-project/sensor-d4-tls-fingerprinting/d4-tlsf.go:354 +0x8b2

**Comment by gallypette** [comment]
Hi, thank you for reporting that I will look into it.

**Comment by kjelle** [comment]
@gallypette, i see that when the ReassembledSG is called for this stream, the stage in the tcpStream's tlsSession is set to 2. So it has seen a ServerHello, but has yet seen the Certificate to finish the stream.

My sensor is running on a large entity, and there is a lot of NATing. Perhaps for some reason the ports (stream indent) is reused, and I see mixed traffic with the same source ip/port - dest ip/port pair?

**Comment by kjelle** [comment]
New update, i see that when this happens the direction of the data is server->client, not client->server.

**Comment by gallypette** [comment]
Try increasing ```MaxBufferedPagesPerConnection``` in d4-tlsf.go l.53 
The page might have been marked for deletion.

Do you have a pcap you can share?

**Comment by kjelle** [comment]
After some debugging, I see that i get a ServerHello in a new stream before, or without, seeing the ClientHello first. Unfortunately the PCAP is from a live system, so I cannot share. I have not tested if I see similar problems in our lab.

**Comment by kjelle** [comment]
@gallypette I think I found the problem for my part.

The problem is that the network traffic delivered to our sensor, or how gopack reassebles the network traffic delivered to our sensor, is our-of-order. As such, I see a variety of combinations in which order the ClientHello(1), ServerHello(2) and Certificate(11) is given to the ReassembledSG.

I replaced your TLSSession.stage with flags and return a function which declares if the TLS Handshake is complete.

I can produce a Merge Request if you want.
